src/webui/api/searchcontroller.cpp

```diff
@@ -0,0 +1,359 @@
+        categories << SearchPluginManager::categoryFullName(category);
+
+    setResult(categories.join(","));
+}
+
+void SearchController::pluginsAction()
+{
+    const QStringList allPlugins = SearchPluginManager::instance()->allPlugins();
+    setResult(getPluginsInfo(allPlugins));
+}
+
+void SearchController::installPluginAction()
+{
+    checkParams({"sources"});
+
+    const QStringList sources = params()["sources"].split('|');
+    for (const QString &source : sources)
+        SearchPluginManager::instance()->installPlugin(source);
+}
+
+void SearchController::uninstallPluginAction()
+{
+    checkParams({"names"});
+
+    const QStringList names = params()["names"].split('|');
+    for (const QString &name : names)
+        SearchPluginManager::instance()->uninstallPlugin(name.trimmed());
+}
+
+void SearchController::enablePluginAction()
+{
+    checkParams({"names", "enable"});
+
+    const QStringList names = params()["names"].split('|');
+    const bool enable = parseBool(params()["enable"].trimmed(), false);
+
+    for (const QString &name : names)
+        SearchPluginManager::instance()->enablePlugin(name.trimmed(), enable);
+}
+
+void SearchController::updatePluginsAction()
+{
+    SearchPluginManager *pluginManager = SearchPluginManager::instance();
+
+    connect(pluginManager, &SearchPluginManager::checkForUpdatesFinished, this, &SearchController::checkForUpdatesFinished);
+    connect(pluginManager, &SearchPluginManager::checkForUpdatesFailed, this, &SearchController::checkForUpdatesFailed);
+    pluginManager->checkForUpdates();
+}
+
+void SearchController::searchTimeout(const QString &sessionId, const QString &id)
+{
+    QHash<QString, SearchHandler *> &searches = m_searches[sessionId];
+
+    if (m_searches.contains(sessionId) && searches.contains(id))
+    {
+        Logger::instance()->addMessage(tr("Deleting search session %1 from web session %2 due to timeout").arg(id).arg(sessionId), Log::INFO);
+        SearchHandler *searchHandler = searches[id];
+        searchHandler->cancelSearch();
+        delete searches.take(id);
+    }
+}
+
+void SearchController::checkForUpdatesFinished(const QHash<QString, PluginVersion> &updateInfo)
+{
+    Logger *const logger = Logger::instance();
+
+    if (updateInfo.isEmpty()) {
+        logger->addMessage(tr("All plugins are already up to date."), Log::INFO);
+        return;
+    }
+
+    logger->addMessage(tr("Updating %1 plugins").arg(updateInfo.size()), Log::INFO);
+
+    SearchPluginManager *pluginManager = SearchPluginManager::instance();
+    for (const QString &pluginName : updateInfo.keys()) {
+        logger->addMessage(tr("Updating plugin %1").arg(pluginName), Log::INFO);
+        pluginManager->updatePlugin(pluginName);
+    }
+}
+
+void SearchController::checkForUpdatesFailed(const QString &reason)
+{
+    Logger::instance()->addMessage(tr("Failed to check for plugin updates: %1").arg(reason), Log::INFO);
```

Comment:
(optional) I would use `LogMsg()` instead of `Logger::instance()->addMessage()` everywhere.